This exercise will require you to pull some data from the Qunadl API. Qaundl is currently the most widely used aggregator of financial market data.

As a first step, you will need to register a free account on the http://www.quandl.com website.

After you register, you will be provided with a unique API key, that you should store:

In [76]:
# Store the API key as a string - according to PEP8, constants are always named in all upper case
API_KEY = 'rvMMg4NrxzuksszYLcxy'

Qaundl has a large number of data sources, but, unfortunately, most of them require a Premium subscription. Still, there are also a good number of free datasets.

For this mini project, we will focus on equities data from the Frankfurt Stock Exhange (FSE), which is available for free. We'll try and analyze the stock prices of a company called Carl Zeiss Meditec, which manufactures tools for eye examinations, as well as medical lasers for laser eye surgery: https://www.zeiss.com/meditec/int/home.html. The company is listed under the stock ticker AFX_X.

You can find the detailed Quandl API instructions here: https://docs.quandl.com/docs/time-series

While there is a dedicated Python package for connecting to the Quandl API, we would prefer that you use the *requests* package, which can be easily downloaded using *pip* or *conda*. You can find the documentation for the package here: http://docs.python-requests.org/en/master/ 

Finally, apart from the *requests* package, you are encouraged to not use any third party Python packages, such as *pandas*, and instead focus on what's available in the Python Standard Library (the *collections* module might come in handy: https://pymotw.com/3/collections/ ).
Also, since you won't have access to DataFrames, you are encouraged to us Python's native data structures - preferably dictionaries, though some questions can also be answered using lists.
You can read more on these data structures here: https://docs.python.org/3/tutorial/datastructures.html

Keep in mind that the JSON responses you will be getting from the API map almost one-to-one to Python's dictionaries. Unfortunately, they can be very nested, so make sure you read up on indexing dictionaries in the documentation provided above.

In [249]:
# First, import the relevant modules
import requests
import json

In [250]:
# Now, call the Quandl API and pull out a small sample of the data (only one day) to get a glimpse
# into the JSON structure that will be returned


url = 'https://www.quandl.com/api/v3/datasets/XFRA/AGB2_UADJ?start_date=2018-11-30&end_date=2018-11-30&api_key=rvMMg4NrxzuksszYLcxy'
r = requests.get(url)
json_data = r.json()

In [251]:
# Inspect the JSON structure of the object you created, and take note of how nested it is,
# as well as the overall structure

# json_data dataset

json_data['dataset']


{'id': 46157305,
 'dataset_code': 'AGB2_UADJ',
 'database_code': 'XFRA',
 'name': 'Agrana Beteiligungs AG (AGB2) Unadjusted Stock Prices',
 'description': ' <b>Ticker</b>: AGB2 <br> <br> <b>Exchange</b>: XFRA <br> <br> Columns: <br> <br> Open,High,Low,Close,Volume are unadjusted and shown in EUR currency. <br> <br> Adjustment Factor shows the factor by which prices are adjusted on days which adjustments take place <br> <br> Adjustment Type is a number representing the type of adjustment. Refer to documentation for more information on these codes.',
 'refreshed_at': '2021-01-02T02:57:19.212Z',
 'newest_available_date': '2018-11-30',
 'oldest_available_date': '2007-01-02',
 'column_names': ['Date',
  'Open',
  'High',
  'Low',
  'Close',
  'Volume',
  'Adjustment Factor',
  'Adjustment Type'],
 'frequency': 'daily',
 'type': 'Time Series',
 'premium': True,
 'limit': None,
 'transform': None,
 'column_index': None,
 'start_date': '2018-11-30',
 'end_date': '2018-11-30',
 'data': [['2018-

In [252]:
json_data['dataset']['data'][0][2]

17.54

These are your tasks for this mini project:

1. Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).
2. Convert the returned JSON object into a Python dictionary.
3. Calculate what the highest and lowest opening prices were for the stock in this period.
4. What was the largest change in any one day (based on High and Low price)?
5. What was the largest change between any two days (based on Closing Price)?
6. What was the average daily trading volume during this year?
7. (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

## 1. Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD)

In [253]:
# Assign URL to variable: new_url
new_url = 'https://www.quandl.com/api/v3/datasets/FSE/AFX_X.json?start_date=2017-01-01&end_date=2017-12-31&api_key=rvMMg4NrxzuksszYLcxy'

#Package the request, send the request and catch the response: new_r
new_r = requests.get(new_url)

## 2. Convert the returned JSON object into a Python dictionary.

In [254]:
#Decode the JSON data into a dictionary: new_json_data
new_json_data = new_r.json()

In [222]:
#new_json_data['dataset']

In [218]:
#new_json_data['dataset']['data']

## 3. Calculate what the highest and lowest opening prices were for the stock in this period.

In [255]:
# list of data including high and low opening prices
list_data = new_json_data['dataset']['data']
 
# Highest opening price during the period of January 01, 2017 to December 31, 2017
highest_price = max([item[2] for item in list_data if item[2] is not None])
print(highest_price)

53.54


In [256]:
# To find the first date of the highest opening price
# Find the first position of item where the value of 'High' is the highest
next(((i, item.index(highest_price))
      for i, item in enumerate(list_data)
      if highest_price in item),
     None)

(9, 2)

In [257]:
list_data[9][0]

'2017-12-14'

In [151]:
#list_data[9]

The highest opening price for the Carl Zeiss Meditec stock on the FSE was **53.54 Euros** in 2017. This occured on **December 14, 2017**

In [258]:
# Lowest opening price during the period of January 01, 2017 to December 31, 2017
lowest_price = min([item[3] for item in list_data if item[3] is not None])
print(lowest_price)

33.62


In [259]:
# To find the first date of the lowest opening price
# Find the first position of item where the value of 'Low' is the lowest
next(((i, item.index(lowest_price))
      for i, item in enumerate(list_data)
      if lowest_price in item),
     None)

(239, 3)

In [260]:
list_data[239][0]

'2017-01-23'

The lowest opening price for the Carl Zeiss Meditec stock on the FSE was **33.62 Euros** in 2017. This occured on **January 23, 2017**

## 4. What was the largest change in any one day (based on High and Low price)?

In [261]:
#Make a list of difference between 'High' and 'Low' on any day
largest_change_list = [item[2] - item[3] for item in list_data]
largest_change = max(largest_change_list)
print(largest_change)

2.8100000000000023


In [262]:
#Find the position of the largest change
largest_change_list.index(largest_change )

161

In [263]:
list_data[161]

['2017-05-11',
 43.4,
 46.06,
 43.25,
 45.0,
 None,
 189125.0,
 8496322.0,
 None,
 None,
 None]

The largest change in opening price in any one day for the Carl Zeiss Meditec stock on the FSE was about **2.81 Euros** in 2017. This occured on **May 11, 2017**

## 5. What was the largest change between any two days (based on Closing Price)?

In [264]:
# list of closing prices
closing_price_list = [item[4] for item in list_data if item[4] is not None]

In [265]:
# largest change between any two days 
largest_change_closing_price =  max(closing_price_list) - min(closing_price_list)
print(largest_change_closing_price)

19.03


In [266]:
closing_price_list.index(max(closing_price_list))

10

In [267]:
list_data[10]

['2017-12-13',
 52.64,
 53.35,
 52.48,
 53.09,
 None,
 128434.0,
 6801159.0,
 None,
 None,
 None]

In [268]:
closing_price_list.index(min(closing_price_list))

239

In [269]:
list_data[239]

['2017-01-23',
 34.04,
 34.12,
 33.62,
 34.06,
 None,
 55333.0,
 1877957.0,
 None,
 None,
 None]

The largest change in closing prices  between any two days was **19.03 euros**, which the difference between when the market experienced the highest closing price of about **53.09 euros** on **December 13, 2017** and when the market experienced the lowest closing price of about **34.06 euros** on **January 23, 2017**.

In [270]:
# list of differences between any two consecutive values
difference_closing_price_list = [abs(closing_price_list[i+1]-closing_price_list[i]) for i in range(len(closing_price_list)-1)]
print(max(difference_closing_price_list))

2.559999999999995


In [271]:
difference_closing_price_list.index(max(difference_closing_price_list))

98

In [272]:
abs(list_data[98][4] - list_data[99][4]) == max(difference_closing_price_list)

True

In [273]:
list_data[98][0]

'2017-08-09'

The chargest change in closing prices between any two consecutive days was about **2.56 euros** which happened between the days of **August 09, 2017** and **August 10, 2017**

## 6. What was the average daily trading volume during this year?

In [274]:
trading_volume_list = [item[6] for item in list_data if item[6] is not None]

In [245]:
def cal_average(num):
    sum_num = 0
    for n in num:
        sum_num = sum_num + n           

    avg = sum_num / len(num)
    return avg

In [275]:
print(cal_average(trading_volume_list))

89124.33725490196


The average daily trading volume during the year 2017 was about **89,124 stocks**.

## 7. (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

In [276]:
def median(L):
    sorted_L = sorted(L)
    len_L = len(L)
    index = (len_L - 1) // 2
   
    if (len_L % 2):
        return sorted_L[index]
    else:
        return (sorted_L[index] + sorted_L[index + 1])/2.0

In [277]:
median(trading_volume_list)

76286.0

The median trading volume during this year was **76,286 stocks**.